File: spoj/classicals/EPALIN.py


def solve(a):
	b = a[::-1]
	i = 0
	j = 0
	while i < len(a):
		if a[i] == b[j]:
			i += 1
			j += 1
		else:
			if a[i] == b[0]:
				j = 0
			else:
				j = 0
				i += 1
	if j < len(a):
		return a+b[j:]
	return a

while 1:
	try:
		x = raw_input().strip()
		if not x: break
		print solve(x)
	except: break

File: spoj/classicals/EPALIN.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

int n;

vector<int> prefix(string s){
	vector<int> p(s.length());
	p[0] = 0;
	for (int i = 1; i < s.length(); i += 1){
		int j = p[i-1];
		while (j > 0 && s[i] != s[j]){
			j = p[j-1];
		}
		if (s[i]==s[j]) p[i] = j+1;
		else p[i] = 0;
	}
	return p;
}

void solve(string s, string ans, string rev){
	vector<int> pre = prefix(s);
	int maxu = pre[pre.size()-1];
	for(int i = maxu; i < n; i += 1){
		ans.push_back(rev[i]);
	}
	cout << ans << endl;;
}

char temp[100001];

int main(void){
	while(scanf("%s",temp) == 1){
		string s = temp;
		string p = "";
		p += s;
		n = s.length();
		string tempo = "";
		reverse(p.begin(),p.end());
		tempo += p;
		p += "#";
		p += s;
		solve(p,s,p);
	}
}